### PR TITLE
add env sources to secrets

### LINKED ISSUE #74: support creating secrets from file contents
For example:

```
secretGenerator:
- name: secret-in-base
  behavior: merge
  commands:
    - type: env
      command: "blackbox_cat secret.env"    
    - type: yaml
      command: "sops -d secret.yaml"    
    - type: json
      command: "cat secret.json"    
    - type: inline
      hello: "echo world"    
```

### PR DESCRIPTION
**Pull request by sethpollack**
fixes #74

### CODE REVIEW

File: pkg/configmapandsecret/secretfactory.go [hunk 1]
```diff
@@ -35,17 +52,73 @@ func (f *SecretFactory) MakeSecret(args types.SecretArgs) (*corev1.Secret, error
...
+		return nil, err
+	}
+	return keyValuesFromLines(content)
+}
+
+func (f *SecretFactory) keyValuesFromCommands(sources map[string]string) ([]kvPair, error) {
+	var kvs []kvPair
+	for k, cmd := range sources {
+		content, err := f.createSecretKey(cmd)
+		fmt.Println("createSecretKey:", content)
```
Comment (Liujingfang1) [inline pkg/configmapandsecret/secretfactory.go]: It's not a good idea to print the secret content. Can we remove this print?

File: pkg/configmapandsecret/secretfactory.go [hunk 2]
```diff
@@ -35,17 +52,73 @@ func (f *SecretFactory) MakeSecret(args types.SecretArgs) (*corev1.Secret, error
...
 	return s, nil
 }
 
+func addKvToSecret(secret *corev1.Secret, keyName, data string) error {
+	// Note, the rules for SecretKeys  keys are the exact same as the ones for ConfigMap.
+	if errs := validation.IsConfigMapKey(keyName); len(errs) != 0 {
+		return fmt.Errorf("%q is not a valid key name for a ConfigMap: %s", keyName, strings.Join(errs, ";"))
+	}
+	if _, entryExists := secret.Data[keyName]; entryExists {
+		return fmt.Errorf("cannot add key %s, another key by that name already exists: %v", keyName, secret.Data)
```
Comment (Liujingfang1) [inline pkg/configmapandsecret/secretfactory.go]: Please not include secret.Data in the error message.

File: pkg/types/kustomization.go [hunk 3]
```diff
@@ -115,11 +115,21 @@ type SecretArgs struct {
 	// keys: "tls.key" and "tls.crt"
 	Type string `json:"type,omitempty" yaml:"type,omitempty"`
 
+	// CommandSources for secret.
+	CommandSources `json:",inline,omitempty" yaml:",inline,omitempty"`
+}
+
+// CommandSources contains some generic sources for secrets.
+// Only one field can be set.
```
[the remark this answers is not in the document]
Comment (sethpollack) [reply pkg/types/kustomization.go]: Sorry, copied that from DataSources. I don't think that is actually the case, I'll remove from both.

File: pkg/configmapandsecret/secretfactory.go [hunk 4]
```diff
@@ -35,17 +52,73 @@ func (f *SecretFactory) MakeSecret(args types.SecretArgs) (*corev1.Secret, error
...
 		}
-		s.Data[k] = out
 	}
+
 	return s, nil
 }
 
+func addKvToSecret(secret *corev1.Secret, keyName, data string) error {
+	// Note, the rules for SecretKeys  keys are the exact same as the ones for ConfigMap.
+	if errs := validation.IsConfigMapKey(keyName); len(errs) != 0 {
```
Comment (monopole) [inline pkg/configmapandsecret/secretfactory.go]: thanks for doing this!

File: pkg/resmap/secret.go [hunk 5]
```diff
@@ -30,7 +30,7 @@ func NewResMapFromSecretArgs(
 	secretList []types.SecretArgs) (ResMap, error) {
 	var allResources []*resource.Resource
 	for _, args := range secretList {
-		s, err := f.MakeSecret(args)
+		s, err := f.MakeSecret(&args)
```
[the remark this answers is not in the document]
Comment (sethpollack) [reply pkg/resmap/secret.go]: I was trying to keep it consistent with MakeConfigMap, should I change that too?

File: pkg/configmapandsecret/secretfactory.go (hunk 1 above)
Comment (sethpollack) [reply pkg/configmapandsecret/secretfactory.go]: oops, that was for debugging :)

File: pkg/resmap/secret.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Liujingfang1) [reply pkg/resmap/secret.go]: I see. Let's keep it consistent with `MakeConfigMap`.

**Comment by lswith** [comment]
can the docs/kustomization.yaml please be updated?

**Pull request merged**